File: benchmarks/subscriber.zig
const std = @import("std");
const nats = @import("nats");
const bench_util = @import("bench_util.zig");

const REPORT_INTERVAL = 10000;

pub const std_options: std.Options = .{
    .log_level = .info,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("Starting NATS subscriber benchmark\n", .{});

    try bench_util.setupSignals();

    // Initialize statistics
    var stats = bench_util.BenchStats.init();

    // Creates a connection to the default NATS URL
    var conn = try bench_util.connect(allocator, null);
    defer bench_util.cleanup(conn);

    const subject = "benchmark.data";

    // Creates a synchronous subscription on subject
    const sub = conn.subscribeSync(subject) catch |err| {
        std.debug.print("Failed to subscribe: {}\n", .{err});
        return err;
    };
    defer {
        conn.unsubscribe(sub) catch {};
        sub.deinit();
    }

    std.debug.print("Subscriber listening on subject '{s}'...\n", .{subject});
    std.debug.print("Press Ctrl+C to stop\n", .{});

    // Wait for messages in a loop
    while (bench_util.keep_running) {
        // Wait for the next message (with timeout)
        var msg = sub.nextMsg(1000) catch break;
        defer msg.deinit();

        stats.msg_count += 1;

        // Print stats every REPORT_INTERVAL messages
        if (stats.msg_count % REPORT_INTERVAL == 0) {
            stats.printThroughput("Received");
        }
    }

    // Print final statistics
    stats.printSummary("messages received");
}

File: benchmarks/subscriber.c
#include "bench_util.h"

#define REPORT_INTERVAL 10000

int main() {
    natsConnection *conn = NULL;
    natsSubscription *sub = NULL;
    natsStatus status;
    bench_stats_t stats;
    
    printf("Starting NATS subscriber benchmark (C/libnats)\n");
    
    // Setup signal handlers
    bench_setup_signals();
    
    // Initialize statistics
    bench_stats_init(&stats);
    
    // Connect to NATS server
    status = bench_connect(&conn, NULL);
    if (status != NATS_OK) {
        return 1;
    }
    
    const char *subject = "benchmark.data";
    
    // Subscribe to subject synchronously
    status = natsConnection_SubscribeSync(&sub, conn, subject);
    if (status != NATS_OK) {
        printf("Failed to subscribe: %s\n", natsStatus_GetText(status));
        bench_cleanup(conn);
        return 2;
    }
    
    printf("Subscriber listening on subject '%s'...\n", subject);
    printf("Press Ctrl+C to stop\n");
    
    // Synchronous message loop
    while (keep_running) {
        natsMsg *msg = NULL;
        
        // Wait for next message with 1 second timeout
        status = natsSubscription_NextMsg(&msg, sub, 1000);
        if (status == NATS_TIMEOUT) {
            continue; // Timeout is expected, just continue
        }
        if (status != NATS_OK) {
            bench_print_error(&stats, status);
            continue;
        }
        
        stats.msg_count++;
        
        // Print stats every REPORT_INTERVAL messages
        if (stats.msg_count % REPORT_INTERVAL == 0) {
            bench_print_throughput(&stats, "Received");
        }
        
        natsMsg_Destroy(msg);
    }
    
    // Print final statistics
    bench_print_summary(&stats, "messages received");
    
    // Cleanup
    natsSubscription_Destroy(sub);
    bench_cleanup(conn);
    
    return 0;
}